### DISCUSSION TITLE
click of a button on pressing enter

### DISCUSSION THREAD

**Discussion by dubbya** [Budibase/budibase#5386] [Help] [answered]
Hi would anyone know how to action a button click from a text box? e.g. if I type something in the text box below and press enter, the lookup button action fires?

Thanks!

![2022-04-13 14_45_25-Window](https://user-images.githubusercontent.com/17187628/163102508-41fee09d-0298-48ec-91c4-830386820525.png)

**Comment by melohagan** [comment]
Hi @dubbya 

There is no way to trigger actions via key-press at the moment. 

More on actions here: https://docs.budibase.com/docs/actions

**Comment by Simoon-F** [comment]
I need

**Comment by melohagan** [comment, marked as answer]
@Simoon-F 
This functionality has since been enabled with **On Change** events! 

 - https://docs.budibase.com/docs/actions#on-change
 - https://docs.budibase.com/docs/text-inputs#on-change

**Reply by Simoon-F** [reply]
![image](https://user-images.githubusercontent.com/48503662/215998527-17eeddb6-6e6a-4d6d-8974-efe314f23cab.png)

I am a novice and don't quite understand. I tried to find the enter key event and did not show that, I may need more help.

**Reply by Simoon-F** [reply]
I want to trigger the OnChange of the button after pressing the enter key.

**Reply by melohagan** [reply]
So the way to think about it is that **Events trigger Actions.**

For example, **On Change** is an event. So if you have a text field with the following:
![Screenshot 2023-02-01 at 09 09 38](https://user-images.githubusercontent.com/101575380/215999816-823559aa-7a9e-405e-8f86-c739cdc43122.png)

That means:
> When the value of the text field changes, then execute all of the actions defined

The text field value change event is triggered when you press enter on a text field (or click off it).

**Reply by Simoon-F** [reply]
![image](https://user-images.githubusercontent.com/48503662/216002705-a115cd11-bd45-4155-a5a8-e2bce1dbbd9f.png)

Thank you for your reply. I have already set the carriage return event of the text box, but how should I trigger the on change event of the button?

**Reply by melohagan** [reply]
Do you mean trigger the on click action of the button? You would simply duplicate the actions found in the **On Click** of the button within the **On Change** of your text field.

**Reply by Simoon-F** [reply]
Thank you for your reply, which is very helpful to me.

**Reply by melohagan** [reply]
You're welcome!

**Reply by NicolasFabre81** [reply]
> Do you mean trigger the on click action of the button? You would simply duplicate the actions found in the **On Click** of the button within the **On Change** of your text field.

Hello @melohagan 
This is not so easy : in case of a SAVE button, when I press ENTER it fires the "onChange" event but as mentionned here https://docs.budibase.com/docs/text-inputs#on-change we need to replace {{ Form.Fields.First Name }} by {{ Field Value }} for the good and current field value to be known

We cannot duplicate actions from the SAVE button to the onchange event...
An if I duplicate manually, I will have 2 places with the same actions, so double work to maintain them...

The optimal would be to have a new action "Click button" that I could call in the OnChange.

**Reply by GeekLad** [reply]
@NicolasFabre81 I'm encountering the same issue.  I have a submit button that performs 10 actions, and I'd like to trigger all those actions when the user presses the enter key in a field.  It's a lot of logic to duplicate, and as you say, if something needs to change it needs to change in both places.  Plus, if you want that same logic in multiple fields (user could hit enter in any field), then it makes it that much more duplicative.

I think a click button action would be a good start to resolve this.  An even better solution would be to add a new feature to allow the builder to create reusable "action sets".  They would behave a lot like methods or functions in coding, and would go a long way to follow DRY (Don't Repeat Yourself) methodology.  It would almost be like automations, but they would be synchronous, and could have bindings just automations can.

**Reply by melohagan** [reply]
Hey @GeekLad 

Not a fan of the click button action because it seems like an anti-pattern / hack to me.

> An even better solution would be to add a new feature to allow the builder to create reusable "action sets". 

I completely agree with this though! 
There is an open issue about re-using code: https://github.com/Budibase/budibase/issues/3571
I've added a comment to suggest allowing action sets to also be re-used. With that said, many actions are context aware (i.e. it depends were the button for example is nested within your component tree) therefore it may be easier said than done.

**Reply by NicolasFabre81** [reply]
Waiting for better, what I did is to put a SAVE button just after the last field.
Users just have to press TAB then ENTER to make the SAVE button fire.
